network/src/gossip.rs

```diff
@@ -129,57 +207,586 @@ pub(crate) struct MessageValidationData {
+		let hash_b = [2u8; 32].into();
+		let hash_c = [3u8; 32].into();
+
+		let message = GossipMessage::Neighbor(VersionedNeighborPacket::V1(NeighborPacket {
+				chain_heads: vec![hash_a, hash_b],
+			})).encode();
+		let res = validator.validate(
+			&mut validator_context,
+			&peer_a,
+			&message[..],
+		);
+
+		match res {
+			GossipValidationResult::Discard => {},
+			_ => panic!("wrong result"),
+		}
+		assert_eq!(
+			validator_context.events,
+			vec![
+				ContextEvent::SendTopic(peer_a.clone(), attestation_topic(hash_a), false),
+				ContextEvent::SendTopic(peer_a.clone(), attestation_topic(hash_b), false),
+			],
+		);
+
+		validator_context.clear();
+
+		let candidate_receipt = CandidateReceipt {
+			parachain_index: 5.into(),
+			collator: [255; 32].unchecked_into(),
+			head_data: HeadData(vec![9, 9, 9]),
+			signature: Default::default(),
+			balance_uploads: Vec::new(),
+			egress_queue_roots: Vec::new(),
+			fees: 1_000_000,
+			block_data_hash: [20u8; 32].into(),
+		};
+
+		let statement = GossipMessage::Statement(GossipStatement {
+			relay_parent: hash_a,
+			signed_statement: SignedStatement {
+				statement: GenericStatement::Candidate(candidate_receipt),
+				signature: Ed25519Signature([255u8; 64]),
+				sender: 1,
+			}
+		});
+		let encoded = statement.encode();
+
+		let topic_a = attestation_topic(hash_a);
+		let topic_b = attestation_topic(hash_b);
+		let topic_c = attestation_topic(hash_c);
+
+		// topic_a is in all 3 views -> succeed
+		validator.inner.write().our_view.add_session(hash_a, MessageValidationData::default());
+		// topic_b is in the neighbor's view but not ours -> fail
+		// topic_c is not in either -> fail
+
+		{
+			let mut message_allowed = validator.message_allowed();
+			assert!(message_allowed(&peer_a, MessageIntent::Broadcast, &topic_a, &encoded));
+			assert!(!message_allowed(&peer_a, MessageIntent::Broadcast, &topic_b, &encoded));
+			assert!(!message_allowed(&peer_a, MessageIntent::Broadcast, &topic_c, &encoded));
+		}
+	}
+
+	#[test]
+	fn too_many_chain_heads_is_report() {
+		let (tx, rx) = mpsc::channel();
+		let tx = Mutex::new(tx);
+		let known_map = HashMap::<Hash, Known>::new();
+		let report_handle = Box::new(move |peer: &PeerId, cb: i32| tx.lock().send((peer.clone(), cb)).unwrap());
+		let validator = MessageValidator::new_test(
+			move |hash: &Hash| known_map.get(hash).map(|x| x.clone()),
+			report_handle,
+		);
+
+		let peer_a = PeerId::random();
+
+		let mut validator_context = MockValidatorContext::default();
+		validator.new_peer(&mut validator_context, &peer_a, Roles::FULL);
+		assert!(validator_context.events.is_empty());
+		validator_context.clear();
+
+		let chain_heads = (0..MAX_CHAIN_HEADS+1).map(|i| [i as u8; 32].into()).collect();
+
+		let message = GossipMessage::Neighbor(VersionedNeighborPacket::V1(NeighborPacket {
+				chain_heads,
+			})).encode();
+		let res = validator.validate(
+			&mut validator_context,
+			&peer_a,
+			&message[..],
+		);
+
+		match res {
```

Comment:
maybe also an `assert_eq!` (with a message)?

Reply:
what message? it shouldn't prompt any send in this test.

Reply:
(if you mean "why use match instead of assert_eq?" the answer is that `GossipValidationResult` is not `PartialEq` or `Debug`)